### DISCUSSION TITLE
Can't write an efficient http to https redirect

### DISCUSSION THREAD

**Discussion by Joaojack** [nuxt/nuxt#16252] [Questions] [answered]
Hello,
I really need your help here. My nuxt 3 website is hosted on heroku, and my SSL certs are ok for my domain in https://domain.com and https://www.domain.com.
But when typing www.domain.com or domain.com, the page is not securized.

In order to try  a http to https redirect I've tried three things :

**The ssl redirect way (version 1)**
After installing redirect-ssl, I try in nuxt config : 
```
import redirectSSL from 'redirect-ssl'
export default ({
(...)
serverMiddleware: [redirectSSL],
(...)
})
```
-> doesn't work

**The ssl redirect way (version2)**
```
const isProd = process.env.NODE_ENV === 'production'od)
const redirectMiddleware = (req: IncomingMessage, res: ServerResponse<IncomingMessage>, next: Function) => {
  const ua = req.headers['user-agent'] || ''
const enabled = isProd && /Mozilla|Chrome|Firefox|Safari/.test(ua)
  if (enabled) {
  return enabled ? redirectSSL(req, res, next) : next()
  }
}
import redirectSSL from 'redirect-ssl'
export default ({
(...)
serverMiddleware: [redirectMiddleware]
(...)
})
```
--> doesn't work

**The ssl redirect way (version3)**
```
import redirectSSL from 'redirect-ssl'
export default ({
(...)
serverMiddleware: redirectSSL.create({enabled: process.env.NODE_ENV === 'production'})],
(...)
})
```
--> doesn't work

** The heroku way**

tried to put this .htaccess file in my root folder : 
```
##Force SSL 

#Normal way (in case you need to deploy to NON-heroku)
RewriteCond %{HTTPS} !=on

#Heroku way
RewriteCond %{HTTP:X-Forwarded-Proto} !https 

#If neither above conditions are met, redirect to https
RewriteRule ^ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]

# If header is present in the request
RewriteCond %{HTTP:X-Forwarded-Proto} .
```
--> doesn't work

**Solution i made up for a temporary but dirty fix**

Forced a hard redirection in my middleware folder when session token is not detected :
```
export default defineNuxtRouteMiddleware(async (to, from) => {
  const myUser = useUser()
  const tokenCookie= useStatefulCookie('tokenCookie')
  if(tokenCookie.value?.access_token)
  { const authorization_middleware=`Bearer ${tokenCookie.value.access_token}`
    if(myUser.value===false&&tokenCookie.value.access_token){
      const user = await $fetch('/api/something',{method:'POST',body:{authorization: authorization_middleware}})
      myUser.value=user
    }}
  else {
    **return navigateTo('https://foobar/login', {external:true})**
}})```

Can some generous soul please help me ? I can't believe setting this redirect is taking me so many time...
thanks

**Comment by wobsoriano** [comment]
Create a server middleware file and use the `fromNodeMiddleware` helper to convert a connect middleware to an h3 event handler:

```ts
// ~/server/middleware/redirect-ssl.ts
import redirectSSL from 'redirect-ssl'

export default fromNodeMiddleware(redirectSSL)
```

**Reply by Joaojack** [reply]
Man, you saved my day, thanks a lot. 
I didn't know this fromNodeMiddleWare helper, i wish I could read more aboutit, but H3 doc doesn't explain anything about it, I would never have found this by myself.

**Reply by TheAlexLichter** [reply]
@Joaojack You can read more in https://github.com/unjs/h3/releases/tag/v0.8.0

**Reply by Joaojack** [reply]
In fact it works in production but my localhost is broken, wether I put in on https with mkcert or not. Do you know how i could exclude localhost from this middleware ?

**Reply by Joaojack** [reply]
in fact, after implementing this solution, a lot of internal server/api calls and root middleware functions were disabled or inoperative, I had to uninstall it and it works again, without the https redirection...
Is there any possibility that this middleware could affect root middleware and api calls ?

**Reply by wobsoriano** [reply]
To exclude localhost, you just need to add it to the `exclude` property of the config

```ts
export default fromNodeMiddleware(redirectSSL.create({ exclude: ['localhost'] }))
```

https://github.com/unjs/redirect-ssl#disable-for-non-production-or-localhost

**Reply by Joaojack** [reply]
unfortunately I've tried it, still error on localhost:3000

**Reply by tadvas** [reply, marked as answer]
That's gonna work

```
import redirectSSL from 'redirect-ssl'

export default fromNodeMiddleware(redirectSSL.create({
  enabled: process.env.NODE_ENV === 'production'
}))
```

**Reply by Joaojack** [reply]
Thanks, tadvas :)
